feat: Display active table filter count #1689

dbranley opened the pull request:
**The PR fulfills these requirements:** (check all the apply)

- [x] It's submitted to the `main` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `feat: Add a button #xxx`, where "xxx" is the issue number).
- [x] When resolving a specific issue, the PR description includes `Closes #xxx`, where "xxx" is the issue number.
- [x] If changes were made to `ui` folder, unit tests (`make test`) still pass.
- [x] New/updated tests are included

Closes #1689

As requested in the ticket, I have updated the table component so that it will display a count of selected options just to the top right of the chevron down icon. As suggested, I used `onRenderHeader` to accomplish this - this means the custom renderer for the column now displays the name, sort icon, chevron icon and count. The code changes are in `ui/src/table.tsx` and here is a screen shot of what it looks like:

![issue1689_output_example](https://github.com/user-attachments/assets/96903b04-f669-43db-9a11-363d5a34733c)

And here is a short video demonstrating the filtering after this change:

https://github.com/user-attachments/assets/fa6f7b85-4a62-40cc-bb08-834b366f4339

Question for you: The count is being rendered in red with a red border to match how the requirement is defined in the ticket. I wasn't sure if that was done in red make it stand-out in the screen shot or if this is how it was supposed to be implemented. Can you confirm if red text with red border is what you expect?

I added 3 new unit tests to `ui/src/table.test.tsx` that check for the new filter counts. These pass along with all the other unit tests. Here is a screen shot of the results from running all the regression tests:

![issue1689_unit-tests_after](https://github.com/user-attachments/assets/f01f33ee-763e-467e-847f-e8fef64cb488)

I also ran the visual regression tool at `tools/showcase`. This was fine except for 3 mismatches found with `time_picker-2.png`, `time_picker-3.png`, and `time_picker-4.png`. I looked at the before and after images and they looked identical to me, so I could not figure out why these were flagged. I suspect these were just false positives?

Let me know how this looks and if you have any questions or suggested changes. Thanks!

mturoci commented:
Thanks @dbranley! You are doing a superb job lately :)

> The count is being rendered in red with a red border to match how the requirement is defined in the ticket. I wasn't sure if that was done in red make it stand-out in the screen shot or if this is how it was supposed to be implemented. Can you confirm if red text with red border is what you expect?

The red was just for illustration. The implementation needs to respect theme colors like the rest of the Wave components. As for the UI design, I would go for badge/chip design, something along the lines of

![image](https://github.com/user-attachments/assets/ac5c527e-3f54-4d7c-be4b-290e1c1bc214)

But more subtle. You can give it a a first go if you want and I can adjust afterwards.

dbranley commented:
@mturoci - Thanks for all the great feedback! I'll get to all this after I'm done with the background image issue. Thanks!

dbranley commented on ui/src/table.test.tsx:
```diff
@@ -1035,14 +1035,84 @@ describe('Table.tsx', () => {
       expect(emitMock).toHaveBeenCalledTimes(1)
     })
+
+    it('Filter counts - manual select, deselect', () => {
```
Done.

dbranley commented on ui/src/table.test.tsx:
```diff
@@ -1035,14 +1035,84 @@ describe('Table.tsx', () => {
       expect(emitMock).toHaveBeenCalledWith(tableProps.name, 'filter', { 'colname2': [] })
       expect(emitMock).toHaveBeenCalledTimes(1)
     })
+
```
Done, new test using 'reset' is added.

dbranley commented on ui/src/table.tsx:
```diff
@@ -335,6 +340,7 @@ const
   DataTable = React.forwardRef(({ model: m, onFilterChange, items, filteredItems, selection, selectedFilters, isMultiple, isSingle, groups, expandedRefs, onSortChange, setFiltersInBulk }: DataTable, ref) => {
     const
       [colContextMenuList, setColContextMenuList] = React.useState<Fluent.IContextualMenuProps | null>(null),
+      [sortCols, setSortCols] = React.useState<Dict<S> | null>(null),
```
I tried using `col.iconName` initially, but it didn't behave as I expected.  All elements of the column header are now handled in `onRenderHeader`, which includes the sorting icon - this is needed to control the placement of each of the elements of the column header. But when `col.iconName` is set, then Fluent tries to render the sort icon, even though it is already defined in `onRenderHeader`. I tried various things to work around this, but no luck. That's why I decided to define a new state variable. Let me know if this makes sense. Thanks!

dbranley replied:
@mturoci, Since I know your busy, I decided to dig a little deeper into your question about why I was unable to use `col.iconName` to hold the state of the sort icon sorting direction. I created a small Fluent app outside of Wave to isolate the situation. It is a new table component that wraps the Fluent `DetailsList` component. The new component is very simple and dumb, with just 3 columns and 5 rows of data. Each column has a ContextMenu that behaves like it would for filtering, but it doesn't actually filter anything. I also keep track of the number of selections for each column and print those at the top.

Here's a screen show of what this looks like:

![issue1689_iconName](https://github.com/user-attachments/assets/3e58b120-4d57-40f3-9f1b-e0fd40a73bca)

Note: You'll find a code snipit where these 3 columns are defined at the bottom of this comment.

The first column is the plainest with no custom renderer used. It has a Sort icon, defined using the `col.iconName` property, the column name, an onColumnClick handler, and a dropdown. The Sort icon doesn't do anything in this example.

The second column uses `onRenderHeader` to display the column header. In onRenderHeader, the column name, Sort icon, filter icon, and selection count are displayed. As with the 1st column, the Sort icon doesn't do anything, just there to display.

The third column is a duplicate of the 2nd column, with the additon of setting `col.iconName`. You'll notice the only thing visible in the 3rd column is the Sort icon. The behavior is the same whether I set this before or after setting the `onRenderHeader` property.

When I looked in the Chrome dev tools, I could see that the `<div>` in onRenderHeader is in the document, but it is on the next line below the header and hidden behind the 1st row of data.

I didn't dig into the Fluent code, but it seems to me that Fluent doesn't behave well if you define a column with both an `iconName` and `onRenderHeader`. And since we need to display the sort icon in `onRenderHeader`, then we need to define a new state variable to hold the sorting direction.

I hope this all makes sense and helps. And if I'm overlooking something, please let me know. Thanks!

Here's the snipit:

```typescript
    const columns: IColumn[] = [
        { key: 'column1', name: 'Name', fieldName: 'name', minWidth: 100, 
          iconName: 'Sort',   
          onColumnClick: (ev, column) => {
            onNameColumnContextMenuHandler(column, ev);
          },
          columnActionsMode: ColumnActionsMode.hasDropdown,
          isResizable: true,
        },        
        { key: 'column2', name: 'Value', fieldName: 'value', minWidth: 100, 
          onRenderHeader: (props?: IDetailsColumnProps) => {
            if (!props) return null
            return (
                <div>
                    {props.column.name}
                    <Icon iconName='Sort'/>
                    <Icon iconName='ChevronDown'
                          onClick={(ev: React.MouseEvent<HTMLElement>) =>{
                            onValueColumnContextMenuHandler(props.column, ev);
                          }}
                    />
                    <span>{valueCheckedCount}</span>
                </div>
            ) 
          },
          isResizable: true, 
        },             
        { key: 'column3', name: 'Company', fieldName: 'company', minWidth: 100, 
            iconName: 'Sort', 
            onRenderHeader: (props?: IDetailsColumnProps) => {
              if (!props) return null
              return (
                  <div>
                      {props.column.name}
                      <Icon iconName='Sort'/>
                      <Icon iconName='ChevronDown'
                            onClick={(ev: React.MouseEvent<HTMLElement>) =>{
                              onCompanyColumnContextMenuHandler(props.column, ev);
                            }}
                      />
                      <span>{companyCheckedCount}</span>
                  </div>
              ) 
            },
            isResizable: true,
            // iconName: 'Sort', 
          }, 
    ];
```

dbranley replied:
@mturoci One more thing, it would be possible to store the state of the sorting direction in `col.iconName`, but that would require 2 things, which I think are a bit of a hack:
1. The `<div>` would need to be defined with `display: 'inline-block'` to prevent the div from starting on the next line - or use a `<span>`; 
2. The Sort icon set with `col.iconName` would have to be defined as both hidden and with a width of 0 so that it wouldn't ever display

The result would be something like this:

![issue1689_sortIcon2](https://github.com/user-attachments/assets/c9e2f5f8-af35-405d-ab85-634c9b131c49)

While this might technically work, it seems more intuitive to me to just define a new state variable for the sorting direction. 

Anyway, wanted to share this. Let me know what you think. thanks!

mturoci replied:
> seems to me that Fluent doesn't behave well if you define a column with both an iconName and onRenderHeader.

That's enough of a reason to me. Didn't check myself, but as long as the new code doesn't cause any regressions, it's fine as is. Thanks for all the digging!

dbranley commented on ui/src/table.tsx:
```diff
@@ -392,10 +390,43 @@ const
+              }
+              {filterCountField ? filterCountField : null}
+              </div> 
+            } else return null
```
Done. Thanks for the suggested refactor - looks a lot better!

dbranley commented:
@mturoci - Thanks for clarifying the requirement for the UI design of the filter count. I removed the border and the red. I also tried to make it have more of a badge design, using a slightly darker background than what is used for the header. Here's a screen shot of what this looks like now:

![issue1689_afterUpdate](https://github.com/user-attachments/assets/57cb4d57-6a20-485a-8bbf-a07375e7f890)

Let me know if this is what you had in mind. 

I also re-ran all the unit tests and everything still passes:

![issue1689_rerunTests](https://github.com/user-attachments/assets/c23f0345-48f4-461b-b5ce-752ddfd0b85c)

I think I've addressed all the review comments. If there's something else you need, let me know. Thanks!

mturoci commented:
Thanks @dbranley! Will try to look into this by the end of the week as I am pretty busy.

dbranley commented:
@mturoci Thanks for catching this! FYI that it does not appear to be a regression, since I was able to replicate the bug even with the original version `table.tsx`, before changes in this PR. But I can still try to fix the bug as part of this PR.

From the debugging I've done so far, I can see that the `menuFilters` state variable within the `ContextualMenu` component is getting updated correctly when `selectAll` is clicked and I have confirmed that the list of `<Fluent.Checkbox>` components contained in `ContextualMenu` are created correctly with the right state for `checked`. But for some reason that I haven't figured out yet, the component is not getting re-rendered consistently, even though the props are getting updated correctly. So far the code in `table.tsx` seems to be doing all the right things, which might mean there is an underlying bug in Fluent. But I've really just started debugging, so it's possible the bug is there in `table.tsx` - I will continue debugging.

Note that since this seems unrelated to this PR, would it make sense to open a new bug for this issue so we can let the PR go through? If so, then I can focus on the bug without holding this PR up. Regardless, I'll keep plugging away at this! Thanks!

mturoci commented:
If it's not a regression, this PR is good to be merged, thanks!
